### ISSUE TITLE
For step2.sh and step3.R

### ISSUE THREAD

**Issue by XRCHAM** [aryeelab/RNAseq_BE_editing#1]
Hi  @sowmyaiyer,

I have a bug when I follow this RNAseq_BE_editing pipeline. After submit the job of step2.sh with 4 options, I got the bug as follows, it seems it could not find "total_reads.treated", which I also didn't find where it was in the script.

Another question is I remember  the paper says "Base edits labeled as C-to-U comprise C-to-U edits called on the positive strand as well as G-to-A edits sourced from the negative strand." While I have no idea how to check the negative or positive strand by script. There seems no specific processes in the script as attached. Would you mind helping me out of these questions?

Thanks,
XR

"gzip: stdin: unexpected end of file

gzip: stdin: unexpected end of file

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

Warning message:
Unknown or uninitialised column: 'total_reads.treated'. 
Error in paste0(REF, "_", ALT) : object 'REF' not found
Calls: %>% ... <Anonymous> -> mutate -> mutate.tbl_df -> mutate_impl -> paste0
Execution halted
srun: error: midway2-0207: task 0: Exited with exit code 1"

**Comment by sowmyaiyer** [comment]
I see a gzip error preceding the error from the R code. 
Could you please upload VCFs and bamreadcounts.gz files and also copy and paste the command you submitted so that I can try to reproduce the error?

**Comment by sowmyaiyer** [comment]
As for your second question, we performed separate analysis to verify that C-U edits were overwhelmingly sourced out of reads from the positive strand whereas G-A edits were coming from the genes expressed from negative strand. This code however, is not part of the standard RNA off-target pipeline.

**Comment by XRCHAM** [comment]
This is my command for step2.sh  and my picture of the vcf file and bamreadcounts.gz file, they are together around 900M. 
(base) [xiangbinruan@midway-login1 comparing_S614vsS2]$ srun -c 1 -n 1 -N 1 -p broadwl-lc --mem-per-cpu=3036M bash ../code/step2.sh --vcf_control ../Output_S210/variantCalls/XZ-XR-1-210_S210_.vcf --readcounts_control ../Output_S210/bam-readcount/XZ-XR-1-210_S210_.all_positions.txt.gz --vcf_treated ../Output_S614/variantCalls/XZ-XR-1-614_S614_.vcf --readcounts_treated ../Output_S614/bam-readcount/XZ-XR-1-614_S614_.all_positions.txt.gz >log 2>err &

**Comment by XRCHAM** [comment]
Thanks for your reply! For the second question, do you mind sharing your method? These methods are foreign to me, I do wet experiments and now I just got some data from Next-Seq, and eager to know whether my experiment works.  
Thanks a lot.

**Comment by sowmyaiyer** [comment]
The usage of the line is with an "=" sign not space.

```
Usage: step2.sh
        --vcf_control=control_vcf_file
        --readcounts_control=control_bam_readcounts_gz
        --vcf_treated=treated_vcf_file
        --readcounts_treated=treated_bam_readcounts_gz
```
Also, if you could upload the actual files on Dropbox or someplace, it would help more than a picture. I understand that there was no error, but I would need those files to reproduce the error you are seeing.

**Comment by XRCHAM** [comment]
Thanks, Sowmyaiyer !  It works when I changed the space to "=".
Cheers!

**Comment by XRCHAM** [comment]
Hi Sowmyaiyer I met problems again about step3.R 
At first I used my whole data and it run a very long time to generate the joined txt file until I cancel(took about 15 hours with 16 threads!) So I subset part of the data by -l option in bam-readcount which I list my mutation sites. Then I  got a very large txt.gz file(15.4 G) after step2, which should be used to plot in R. When running step3.R , at first it indicates out of memory, so I gave a large mem(120G). But it indicate another error(see below). Finally I only subset the first 100 rows of each vcf file and according to the new vcf file, I generated new bamreadcount txt file, much smaller txt which are attached below. Now it still give the same err as the 15.4G file.

Could you help me out of this problem ?

**Error like this:**

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

Parsed with column specification:
cols(
  .default = col_double(),
  poskey = col_character(),
  POS = col_character(),
  other = col_logical(),
  REF = col_character(),
  QUAL = col_character(),
  N.treated = col_character(),
  G.control = col_character(),
  T.control = col_character(),
  N.control = col_character()
)
See spec(...) for full column specifications.
Warning: 1370000 parsing failures.
row col   expected     actual                                                                      file
  1  -- 23 columns 20 columns 'XZ-XR-1-614_S614_AG-100.vcf.sorted.vcf_with_coverage.be_and_control.txt'
  2  -- 23 columns 20 columns 'XZ-XR-1-614_S614_AG-100.vcf.sorted.vcf_with_coverage.be_and_control.txt'
  3  -- 23 columns 20 columns 'XZ-XR-1-614_S614_AG-100.vcf.sorted.vcf_with_coverage.be_and_control.txt'
  4  -- 23 columns 20 columns 'XZ-XR-1-614_S614_AG-100.vcf.sorted.vcf_with_coverage.be_and_control.txt'
  5  -- 23 columns 20 columns 'XZ-XR-1-614_S614_AG-100.vcf.sorted.vcf_with_coverage.be_and_control.txt'
... ... .......... .......... .........................................................................
See problems(...) for more details.

Error in G.control/total_reads.control : 
  non-numeric argument to binary operator
Calls: %>% ... mutate.tbl_df -> mutate_impl -> case_when -> eval_tidy
Execution halted
srun: error: midway2-0057: task 0: Exited with exit code 1
err (END) 

**The command I used as follow, which I only paste the control group(XZ-XR-1-210_S210_AG-100).**

#subset the first 100 row of vcf file 
head -100 XZ-XR-1-210_S210_AG.vcf > XZ-XR-1-210_S210_AG-100.vcf

#generate first 100 site list for bam-readcount
cat XZ-XR-1-210_S210_AG-100.vcf|tail -n +3403 |awk '{FS="\t";OFS="\t";print $1,$2,$2}' >XZ-XR-1-210_S210_AG-100.list &

#generate new txt file for the first 100 sites
bam-readcount -w 1 -l XZ-XR-1-210_S210_AG-100.list -f /project2/xczhang/genomes/hg38/Homo_sapiens_assembly38.fasta ../Output_S210/forVariantCalling/XZ-XR-1-210_S210_.bam  | sed 's/:/\t/g' | awk 'BEGIN{print "CHR\tBP\tREF\tTOTAL\tA\tC\tG\tT\tN"}; {print $1,$2,$3,$4,$20,$34,$48,$62,$76}' OFS='\t' | gzip > XZ-XR-1-210_S210_AG-100.txt.gz 2>err &

#the same command was done to treated group, and run step2/step3
srun -c 2 -n 1 -N 1 -p broadwl --mem-per-cpu=10000M bash ../code/step2.sh \
--vcf_control=XZ-XR-1-210_S210_AG-100.vcf \
--readcounts_control=XZ-XR-1-210_S210_AG-100.txt.gz \
--vcf_treated=XZ-XR-1-614_S614_AG-100.vcf \
--readcounts_treated=XZ-XR-1-614_S614_AG-100.txt.gz \
>log 2>err &

**Files I used for step2, for the vcf files , I compressed with bgzip for uploading**

[XZ-XR-1-210_S210_AG-100.txt.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195246/XZ-XR-1-210_S210_AG-100.txt.gz)

[XZ-XR-1-614_S614_AG-100.txt.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195247/XZ-XR-1-614_S614_AG-100.txt.gz)

[XZ-XR-1-210_S210_AG-100.vcf.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195248/XZ-XR-1-210_S210_AG-100.vcf.gz)
[XZ-XR-1-614_S614_AG-100.txt.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195249/XZ-XR-1-614_S614_AG-100.txt.gz)

**Comment by XRCHAM** [comment]
I am sorry, the loading files above are something wrong, these are new ones:
[XZ-XR-1-210_S210_AG-100.txt.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195260/XZ-XR-1-210_S210_AG-100.txt.gz)
[XZ-XR-1-210_S210_AG-100.vcf.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195261/XZ-XR-1-210_S210_AG-100.vcf.gz)
[XZ-XR-1-614_S614_AG-100.txt.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195262/XZ-XR-1-614_S614_AG-100.txt.gz)
[XZ-XR-1-614_S614_AG-100.vcf.gz](https://github.com/aryeelab/RNAseq_BE_editing/files/3195263/XZ-XR-1-614_S614_AG-100.vcf.gz)

**Comment by sowmyaiyer** [comment]
Here is what I propose.
Split your bamreadcounts and vcfs by chromosome like this

```
# split bam readcounts by chr
zcat ${bamreadcountfile} | awk -vtemp_folder=${temp_folder} '{ print $1"_"$2"\t"$0 >> temp_folder"/chr/"`basename bamreadcountfile`".all_positions."$1".txt"}'
# split vcf file by chr
sed '/##/d' ${vcffile} | awk -vtemp_folder=${temp_folder} '{ print $1"_"$2"\t"$0 >> temp_folder/chr/`basename bamreadcountfile`"."$1".vcf"}'
```

gzip the individual bamreadcount files to be ready for step2
```
for chr_split_file in ${temp_folder}/chr/`basename bamreadcountfile`.all_positions.*.txt
do
        gzip ${chr_split_file}
        <submit as job> ./step2.sh  --vcf_control=control_vcf_file_chr  --readcounts_control=control_bam_readcounts_gz_chr --vcf_treated=treated_vcf_file_chr --readcounts_treated=treated_bam_readcounts_gz_chr
done
```

#### Now call step2.sh on the chromosome-wise split files for each sample and each chromosome. This is the memory-intensive step and can be run in parallel for all chromosomes. Make sure you name the outputs by sample so that you can concatenate by sample after all jobs are done. ####

```
<submit as job> ./step2.sh  --vcf_control=control_vcf_file_chr  --readcounts_control=control_bam_readcounts_gz_chr --vcf_treated=treated_vcf_file_chr --readcounts_treated=treated_bam_readcounts_gz_chr
```

##### combine outputs by sample, from step2 using cat
```
cat samplename.chr*.sorted.vcf_with_coverage.be_and_control.txt > samplename.vcf_with_coverage.be_and_control.txt
```

#### call step3 on the concatenated output above ####
```
Rscript step3.R samplename.sorted.vcf_with_coverage.be_and_control.txt
```

**Comment by sowmyaiyer** [comment]
I do see a step that I had missed including when I published the code. I will put out a fix shortly. In the meantime, the above code for splitting by chromosome should work since that includes the missing code.

**Comment by sowmyaiyer** [comment]
I was able to reproduce your bug, put in a bug fix and I ran it with the four files you provided and the code ran with no errors.

**Comment by XRCHAM** [comment]
Thanks, Sowmyaiyer!  It is very nice of you to feedback in time. I will follow what you have proposed. 

Cheers,
Xiangbin

**Comment by sowmyaiyer** [comment]
Hi Xiangbin,

Were you able to run the code? If so, could you please close the issue?
If you are still having trouble, please let me know.

Sowmya

**Comment by XRCHAM** [comment]
acutally I met almost the same problems as last time using the new script, still could not run over the process. Now I am trying other methods. Thanks any way.
